src/zalsa.rs

```diff
@@ -77,9 +77,14 @@ static NONCE: NonceGenerator<StorageNonce> = NonceGenerator::new();
 pub struct IngredientIndex(u32);
```

Comment:
Could we use a `NonZeroU32` instead of a tag? or does that not work because `QueryEdge` is wrapping more than just `IngredientIndex`.

Reply:
I think the main annoyance here comes with the fact that we would want a `NonMax` variant, and so we'd have to do the -1 / +1 dance on use / construction then. I wonder if we can just put a `NonZeroU32` here and allocate an unused dummy ingredient into slot 0 to avoid having to do that?

Reply:
The issue is that we need to reserve a whole bit for the tag, not just a niche (so this cuts the range of indices in half).